tests/unit/test_charm.py

```diff
@@ -17,72 +17,105 @@
+        self.mock_nrf_url = TestCharm.patcher_nrf_url.start()
+        self.mock_request_certificate_creation = TestCharm.patcher_request_cert_creation.start()
+        self.mock_home_network_private_key = X25519PrivateKey.generate()
+        self.mock_cryptography_generate = TestCharm.patcher_cryptography_generate.start()
+        self.mock_cryptography_generate.return_value = self.mock_home_network_private_key
+        metadata = self._get_metadata()
+        self.container_name = list(metadata["containers"].keys())[0]
+
+    def tearDown(self) -> None:
+        patch.stopall()
+
+    @pytest.fixture()
+    def mock_default_values(self):
+        self.mock_nrf_url.return_value = VALID_NRF_URL
+        self.mock_check_output.return_value = POD_IP
+        self.mock_generate_private_key.return_value = PRIVATE_KEY.encode()
+        self.mock_generate_csr.return_value = CSR.encode()
+
+    @pytest.fixture(autouse=True)
+    def harness(self, setUp, request, mock_default_values):
         self.harness = testing.Harness(UDMOperatorCharm)
-        self.harness.set_model_name(name=self.namespace)
-        self.addCleanup(self.harness.cleanup)
+        self.harness.set_model_name(name=NAMESPACE)
         self.harness.set_leader(is_leader=True)
         self.harness.begin()
+        yield self.harness
+        self.harness.cleanup()
+        request.addfinalizer(self.tearDown)
+
+    @pytest.fixture()
+    def add_storage(self):
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.add_storage(storage_name="config", attach=True)
 
     @staticmethod
     def _get_metadata() -> dict:
-        """Read `charmcraft.yaml` and returns it as a dictionary.
-
-        Returns:
-            dics: charmcraft.yaml as a dictionary.
-        """
+        """Read `charmcraft.yaml` and return its content as a dictionary."""
         with open("charmcraft.yaml", "r") as f:
             data = yaml.safe_load(f)
         return data
 
     @staticmethod
     def _read_file(path: str) -> str:
-        """Read a file and returns as a string.
-
-        Args:
-            path (str): path to the file.
-
-        Returns:
-            str: content of the file.
-        """
+        """Read a file from a given path and return its content as a string."""
         with open(path, "r") as f:
             content = f.read()
         return content
 
-    def _create_nrf_relation(self) -> int:
-        """Create NRF relation.
+    @staticmethod
+    def _get_provider_certificate():
+        provider_certificate = Mock(ProviderCertificate)
+        provider_certificate.certificate = CERTIFICATE
+        provider_certificate.csr = CSR
+        return provider_certificate
 
-        Returns:
-            int: relation id.
-        """
+    def _create_nrf_relation(self) -> int:
+        """Create NRF relation and return its relation id."""
         relation_id = self.harness.add_relation(
```

Comment:
```suggestion
        relation_id = self.harness.add_relation(  # type:ignore

```

Fails in static check because of attribute error.